python/mxnet/_ctypes/ndarray.py

```diff
@@ -93,13 +101,28 @@ def _imperative_invoke(handle, ndargs, keys, vals, out):
 
     if original_output is not None:
         return original_output
-    if num_output.value == 1:
-        return _ndarray_cls(ctypes.cast(output_vars[0], NDArrayHandle),
-                            stype=out_stypes[0])
+    if _is_np_op(handle):
+        if num_output.value == 1:
+            return _np_ndarray_cls(ctypes.cast(output_vars[0], NDArrayHandle))
+        else:
+            return [_np_ndarray_cls(ctypes.cast(output_vars[i], NDArrayHandle))
+                    for i in range(num_output.value)]
     else:
-        return [_ndarray_cls(ctypes.cast(output_vars[i], NDArrayHandle),
-                             stype=out_stypes[i])
-                for i in range(num_output.value)]
+        if num_output.value == 1:
+            return _ndarray_cls(ctypes.cast(output_vars[0], NDArrayHandle),
+                                stype=out_stypes[0])
+        else:
+            return [_ndarray_cls(ctypes.cast(output_vars[i], NDArrayHandle),
+                                 stype=out_stypes[i])
+                    for i in range(num_output.value)]
+
+
+def _is_output_from_np_op(handle, idx):
+    """Check if the CachedOp's idx-th output is from a numpy op."""
```

Comment:
if this is only for cached op, should it be made part of CachedOp class?

Reply:
Done.